### PR TITLE
refactor/softmax

### PR DESCRIPTION
**Pull request by iainwo**
Was just reading through the library and trying to understand it. Made some changes that might/might-not be worth it! Let me know what you guys think.

Generalized logsoftmax and sigmoid with softmax.

- 0.3% less lines 😊 
- fewer calculations
    - we need to calculate the gradient of logsoftmax in backward pass
    - previous impl does this by exponentiating the forward pass output -> this means the exp of a log of a exp
    - instead we can remove one of the exponentiation ops by going straight out of logspace
        - by calculating the softmax
            - taking the log for the forward pass
            - reusing the softmax in the backward pass -> since softmax is the derivative of logsumexp
- more numerically stable (log of softmax vs exp of logsumexp)
- maybe bad because more memory for sigmoid padding

References:
- https://math.stackexchange.com/a/2340848
- https://timvieira.github.io/blog/post/2014/02/11/exp-normalize-trick/
- https://blog.feedly.com/tricks-of-the-trade-logsumexp/

### CODE REVIEW

**Comment by geohot** [comment]
So I def like it in logsoftmax, I'm confused about it in sigmoid though

**Comment by kartik4949** [comment]
> So I def like it in logsoftmax, I'm confused about it in sigmoid though. I'm not even sure we need the np.where in sigmoid.

same

**Comment by iainwo** [comment]
Hey both,

Can remove sigmoid code from PR!

Here were notes/thoughts I've come up for sigmoid. Give it a read - and let's decide whether to keep it, remove it (back to previous impl) or replace it (with np.piecewise).

- sigmoid is 2-class variant of softmax (https://stats.stackexchange.com/a/254071)
- either mathematical formulation of sigmoid are computationally unstable (https://timvieira.github.io/blog/post/2014/02/11/exp-normalize-trick/)
    -  overflow possibility: `x -> inf` and `sigmoid(x) -> exp(x)/(1 + exp(x))`
    -  overflow possibility: `x -> -inf` and `sigmoid(x) -> 1/(1 + exp(-x))`
    - takeaway, need to guard when `abs(x) -> inf`
- ways of guarding sigmoid stability
    1. np.where
    2. zero padding + a zero-normalized exponential (_exp_normalize)
        - we can shift the domain of the input since sigmoid is phase invariant
        - that is: sigmoid([0, x]) == sigmoid([-x, 0]])
    3. np.piecewise (https://stackoverflow.com/a/62860170)
        - big con: only works for floats
- i'm thinking it's possible for tensor inputs to be arbitrarily big/small, which means we need to guard against overflows in sigmoid. I know you guys mentioned removing np.where - maybe theres an another solution?

Iain

**Comment by iainwo** [comment]
removed sigmoid code - since it seems like a trivial objection. can open another with one of the three aforementioned options if we think it's worth it.

There's probably more interesting/useful things to do and work on with tinygrad 😁

**Comment by geohot** [comment]
The padding is what confuses me. Like is sigmoid([0, x]) == sigmoid([-x, 0]]) really true?

```
>>> import numpy as np
>>> sigmoid = lambda input: 1/(1 + np.exp(-np.array(input)))
>>> sigmoid([0,5]), sigmoid([-5,0])
(array([0.5       , 0.99330715]), array([0.00669285, 0.5       ]))
```

**Comment by kartik4949** [comment]
also  if it's so confusing, that's a sign of bad readability

**Comment by geohot** [comment]
I mean, the np.where isn't great either.

**Comment by kartik4949** [comment]
hmm... well ill work to make it great then!, 'may be refering pytorch c/c++ code.'

**Comment by iainwo** [comment]
> The padding is what confuses me. Like is sigmoid([0, x]) == sigmoid([-x, 0]]) really true?
> 
> ```
> >>> import numpy as np
> >>> sigmoid = lambda input: 1/(1 + np.exp(-np.array(input)))
> >>> sigmoid([0,5]), sigmoid([-5,0])
> (array([0.5       , 0.99330715]), array([0.00669285, 0.5       ]))
> ```

@geohot - yeah your absolutely right miswrote that in comment - in the code i was padding and calling softmax (not sigmoid)!
- sigmoid implicitly is a softmax. so `sigmoid([x])` -> `[softmax([0,x]) for x in tensor_vals]`

Here is an example:
```
>>> import torch
>>> tenz = torch.Tensor([0, 5])

# softmax invariance
>>> torch.nn.Softmax(dim=0)(tenz)
tensor([0.0067, 0.9933])
>>> torch.nn.Softmax(dim=0)(tenz-5)
tensor([0.0067, 0.9933])
```
